android-core/src/main/java/com/mparticle/internal/DeviceAttributes.kt

```diff
@@ -0,0 +1,343 @@
+            attributes.put(MessageKey.TIMEZONE, TimeZone.getDefault().rawOffset / (1000 * 60 * 60))
+            // network
+            val telephonyManager = appContext
+                .getSystemService(Context.TELEPHONY_SERVICE) as TelephonyManager
+            val phoneType = telephonyManager.phoneType
+            if (phoneType != TelephonyManager.PHONE_TYPE_NONE) {
+                // NOTE: network attributes can be empty if phone is in airplane
+                // mode and will not be set
+                val networkCarrier = telephonyManager.networkOperatorName
+                if (0 != networkCarrier.length) {
+                    attributes.put(MessageKey.NETWORK_CARRIER, networkCarrier)
+                }
+                val networkCountry = telephonyManager.networkCountryIso
+                if (0 != networkCountry.length) {
+                    attributes.put(MessageKey.NETWORK_COUNTRY, networkCountry)
+                }
+                // android combines MNC+MCC into network operator
+                val networkOperator = telephonyManager.networkOperator
+                if (6 == networkOperator.length) {
+                    attributes.put(MessageKey.MOBILE_COUNTRY_CODE, networkOperator.substring(0, 3))
+                    attributes.put(MessageKey.MOBILE_NETWORK_CODE, networkOperator.substring(3))
+                }
+            }
+            attributes.put(MessageKey.DEVICE_IS_TABLET, MPUtility.isTablet(appContext))
+            attributes.put(MessageKey.DEVICE_IS_IN_DST, MPUtility.isInDaylightSavings())
+
+            if (!MPUtility.isEmpty(deviceImei)) {
+                attributes.put(MessageKey.DEVICE_IMEI, deviceImei)
+            }
+        } catch (e: Exception) {
+            // believe it or not, difference devices can be missing build.prop fields, or have otherwise
+            // strange version/builds of Android that cause unpredictable behavior
+        }
+
+        return attributes
+    }
+
+    /**
+     * For the following fields we always want the latest values
+     */
+    fun updateDeviceInfo(context: Context, deviceInfo: JSONObject) {
+        deviceInfo.remove(MessageKey.LIMIT_AD_TRACKING)
+        deviceInfo.remove(MessageKey.GOOGLE_ADV_ID)
+        val adIdInfo = MPUtility.getAdIdInfo(context)
+        var message =
+            "Failed to collect Advertising ID, be sure to add Google Play services (com.google.android.gms:play-services-ads) or Amazon Ads (com.amazon.android:mobile-ads) to your app's dependencies."
+        if (adIdInfo != null) {
+            try {
+                deviceInfo.put(MessageKey.LIMIT_AD_TRACKING, adIdInfo.isLimitAdTrackingEnabled)
+                val instance = MParticle.getInstance()
+                // check instance nullability here and decline to act if it is not available. Don't want to have the case where we are overriding isLimiAdTrackingEnabled
+                // just because there was a timing issue with the singleton
+                if (instance != null) {
```

Comment:
Same question here about the null check vs the let syntax. Mainly trying to see if we can standardize on just try (assuming that's idiomatic).

Reply:
//check instance nullability here and decline to act if it is not available. Don't want to have the case where we are overriding isLimiAdTrackingEnabled
                //just because there was a timing issue with the singleton
                
The previous developer added this comment for this condition.